napari/_vispy/camera.py

```diff
@@ -133,11 +136,33 @@ def perspective(self, perspective):
         self._3D_camera.fov = perspective
         self._view.camera.view_changed()
 
+    @property
+    def mouse_zoom(self):
+        return self._view.camera.mouse_zoom
+
+    @mouse_zoom.setter
+    def mouse_zoom(self, mouse_zoom):
+        self._view.camera.mouse_zoom = mouse_zoom
+
+    @property
+    def mouse_pan(self):
+        return self._view.camera.mouse_pan
+
+    @mouse_pan.setter
+    def mouse_pan(self, mouse_pan):
+        self._view.camera.mouse_pan = mouse_pan
+
     def _on_ndisplay_change(self):
+        mouse_pan, mouse_zoom = self.mouse_pan, self.mouse_zoom
+
         if self._dims.ndisplay == 3:
             self._view.camera = self._3D_camera
         else:
             self._view.camera = self._2D_camera
+
+        # Preserve the mouse behaviour between the cameras
+        self.mouse_pan, self.mouse_zoom = mouse_pan, mouse_zoom
+
```

Comment:
I'm confused about why this is needed... Does updating `self._view.camera` really propagate changes up two levels to change these values?

Reply:
(If so, then maybe capture that in a comment to explain why these things are here?)

Reply:
These lines are needed to keep these states consistent between two different cameras (2D and 3D). The current implementation of `_vispy/camera.Camera` actually creates and keeps two independent instances of cameras at the same time that can be switched during runtime.

>  Does updating self._view.camera really propagate changes up two levels to change these values?
There is no any propagation, there are only two instances of cameras `self._2D_camera`, `self._3D_camera`. `self._view.camera` is just a reference to them.

Reply:
And if you remove these lines, `mouse_pan`, `mouse_zoom` will live independently inside each camera

Reply:
Actually, the design of this class is not very good, there is lots of code that is actually needed only to keep states synchronized between cameras (in an indirect way)

Reply:
I refactored that part so that it is incorporated into this class more naturally: https://github.com/napari/napari/pull/5701/commits/013218f31fe5fa0bc008ecff669d4235ff4c15f9